### PR TITLE
Anmol/release static

### PR DESCRIPTION
**Pull request by Anmol1696**
## 1. Overview

Create static release binaries in actions via building in docker.
Combine build and publish into one step.
Inspiration: https://github.com/osmosis-labs/osmosis/blob/main/.github/workflows/release.yml

## 2. Implementation details
* Create `Dockerfile.release` which is used for release
* Move dockerfiles in `docker` dir
* Add make commands for release yaml
* Rename `deploy.yml` to `release.yml`
* Build based on platform, instead of having 2 commands for 2 platforms, we give `PLATFORM` as variable

## 3. How to test/use
* Run `make release-build-platform PLATFORM=amd64 VERSION=test`
* Run `make release-build-platform PLATFORM=arm64 VERSION=test`
* Run `make release-git release-sha`
* Check the `release/` dir

## 4. Checklist
- [ ] Does the Readme need to be updated?

## 5. Limitations (optional)
* For buiilding multi-platform binaries, need to call the make command 2 times
* Additional `docker/Dockerfile.release` dockerfile, maybe we need to find a way to combine the 2

## 6. Future Work (optional)
* Combine docker files into more unified images

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  release:
+    types: [ published ]
+
+# A workflow run is made up of one or more jobs that can run sequentially or in parallel
+jobs:
+  release:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check out the repo
+        uses: actions/checkout@v2
```
[the remark this answers is not in the document]
Comment (Anmol1696) [reply .github/workflows/release.yml:15]: mhh does not matter much, but sure can change it

Comment (Anmol1696) [inline .github/workflows/release.yml:15]:
```suggestion
        uses: actions/checkout@v3
```

**Review by puneet2019 — APPROVED** [review]

File: Makefile [hunk 2]
```diff
@@ -199,3 +178,40 @@ docker-clean-image:
...
+	tar -zcvf release/persistenceCore-$(VERSION)-linux-$(PLATFORM).tar.gz release/persistenceCore-$(VERSION)-linux-$(PLATFORM)
+	-@$(DOCKER) rm -f release-$(PLATFORM)
+
+release-sha:
+	mkdir -p release/
+	rm -f release/sha256sum.txt
+	sha256sum release/* | sed 's#release/##g' > release/sha256sum.txt
+
+# Create git archive
+release-git:
```
Comment (snj33v) [inline Makefile:204]: github does this for us

File: Makefile [hunk 3]
```diff
@@ -108,24 +109,9 @@ $(BUILD_TARGETS): go.sum $(BUILDDIR)/
 $(BUILDDIR)/:
 	mkdir -p $(BUILDDIR)/
 
-build-reproducible: go.sum
```
Comment (snj33v) [inline Makefile:111]: this really useful have a command to do reproducible builds

Comment (snj33v) [reply Makefile:111]: we also need  to expand on this based on
https://github.com/cosmos/gaia/pull/523

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -0,0 +1,30 @@
+name: Release
+
+# Controls when the action will run. Triggers the workflow on push or pull request
+# events but only for the master branch
+on:
+  release:
+    types: [ published ]
+
+# A workflow run is made up of one or more jobs that can run sequentially or in parallel
```
Comment (snj33v) [inline .github/workflows/release.yml:9]: please use runs-on: ${{ matrix.config.os }}.
and create builds for arch using https://github.com/marketplace/actions/run-on-architecture

**Review by snj33v — CHANGES_REQUESTED** [review]
our discussion on building static binaries aside, we need to orchestrate GitHub actions effectively,

**Comment by Anmol1696** [comment]
@snj33v want to open this PR again, after our talk. 
As we discussed we will add a separate feature request that will give validators option to have shared library non static binaries as well.
For now we will keep releasing new binaries when the underlying c libraries update

**Review by snj33v — APPROVED** [review]

**Pull request merged**